src/com/amazon/ion/impl/lite/IonValueLite.java

```diff
@@ -1002,19 +985,59 @@ final void writeTo(IonWriter writer, SymbolTableProvider symbolTableProvider)
             writer.setFieldNameSymbol(tok);
         }
 
-        SymbolToken[] annotations = getTypeAnnotationSymbols(symbolTableProvider);
+        SymbolToken[] annotations = value.getTypeAnnotationSymbols(symbolTableProvider);
         writer.setTypeAnnotationSymbols(annotations);
+    }
 
-        try
-        {
-            writeBodyTo(writer, symbolTableProvider);
-        }
-        catch (IOException e)
-        {
+    private void writeToIterative(IonWriter writer, SymbolTableProvider symbolTableProvider) throws IOException {
+        Deque<Iterator<IonValue>> iteratorStack = null;
+        Iterator<IonValue> currentIterator = null;
+        IonValueLite value = this;
+        do {
+            writeFieldNameAndAnnotations(writer, value, symbolTableProvider);
+            IonType type = value.getType();
+            if (value.isNullValue()) {
+                writer.writeNull(type);
+            } else if (IonType.isContainer(type)) {
+                if (currentIterator != null) {
+                    if (iteratorStack == null) {
+                        iteratorStack = new ArrayDeque<>(CONTAINER_STACK_INITIAL_CAPACITY);
+                    }
+                    iteratorStack.add(currentIterator);
+                }
+                currentIterator = ((IonContainer) value).iterator();
+                writer.stepIn(type);
+            } else {
+                value.writeBodyTo(writer, symbolTableProvider);
+            }
```

Comment:
Your mileage may vary, but when I have multiple simple cases and a complex case I prefer to put the simple cases together, I find it easier to read that way. 

```suggestion
            if (value.isNullValue()) {
                writer.writeNull(type);
            } else if (!IonType.isContainer(type)) {
                value.writeBodyTo(writer, symbolTableProvider);
            } else { // IonType.isContainer(type)
                if (currentIterator != null) {
                    if (iteratorStack == null) {
                        iteratorStack = new ArrayDeque<>(CONTAINER_STACK_INITIAL_CAPACITY);
                    }
                    iteratorStack.add(currentIterator);
                }
                currentIterator = ((IonContainer) value).iterator();
                writer.stepIn(type);
            }
```